### ISSUE TITLE
Symbol lookup is PAINFULLY SLOW

### ISSUE THREAD

**Issue by LeadAssimilator** [microsoftfeedback/WinDbg-Feedback#131]
Try debugging / loading a huge app with pdbs in the gb range (take edge or chrome for example), and search or fat finger a symbol name, and it takes forever to resolve.

**Comment by KenSykes** [comment]
Yes, this is a problem, and we are working on faster download of large PDBs.

**Comment by LeadAssimilator** [comment]
@KenSykes The issue isn't about symbol download, it is about symbol lookup/search performance. And secondarily, how easy it is to accidentally type the wrong symbol in say the disassembly view and that one has to wait an extremely long time for a previous lookup to complete before the corrected lookup is even attempted - i.e. previous lookups should at the very least be immediately cancelable via some sort of break sequence and/or automatically canceled and restarted after correcting the symbol and pressing enter in the various views that accept symbol names, such as the disassembly view.

**Comment by KenSykes** [comment]
The primary slowdown for symbol lookup is the need to download the PDB and the larger the PDB, the bigger the penalty. That said, I also understand the request to cancel these operations, this is why the issue is still open. This has been requested for a long time and has not risen to enough importance to fix relative to all of our other priorities unfortunately.

As for pure lookup speed there are some known issues around expensive regex searches, I could see those potentially getting fixed one day. There *is* a negative cache to remember symbols that are not found, but that won't help accidental searches.

**Comment by LeadAssimilator** [comment]
Ok, but just to clarify again for the purpose of this specific issue, I'm specifically citing the poor performance of symbol lookup only for large, already fully downloaded and loaded pdbs, which is distinct and separate from any performance problems that exist downloading or loading pdbs.

Given the size of some pdbs, I think it is reasonable to wait for download and load to complete, even if it is somewhat slow. Given the current size of Edge or Chrome pdbs, I don't find the load wait unreasonable, which is why I'm trying to clarify the scope of this issue and limit it to only lookup performance and not anything else. 

To put it in perspective, for projects with huge pdbs, their compile/link step wait is massive anyhow (at least an order more), so an additional large pdb load penalty on start or first symbol lookup is acceptable. However, for complex debug only workflows, fast symbol lookup is important, and the current poor performance makes everything take an unreasonably long amount of time.

If/when this issue is scheduled to be addressed internally, someone on the Edge team would be a good point of contact for determining what the target metrics should be.

**Comment by yharaskrik42** [comment]
I want to throw in my observations (using WinDbgX 1.2601.12001.0). I have attached a debugger to the `EngHost.exe` process while waiting VERY long for `!analyze` to finish.
The time is spend in `dbghelp.dll!FindExecutableImageExW` which does an `EnumDirTreeW` of the symbol directory(s) - our symbol directory lies on the network and is quite large. So especially for not present DLLs the search takes VERY long.
A look into the Process Monitor also shows that every subdirectory is searched for the missing DLL (in our case a Native Interface DLL).
Changing the symbol path to only contain local paths (with SSDs) speeds up the search.

**Comment by KenSykes** [comment]
@yharaskrik42  - A common mistake is pointing to a network share of symbols without specifying srv*. If you do that the debugger will indeed search the directory for PDBs. If the network share is setup as a symbol server, it will have a pingme.txt in the root and a directory for each pdb/binary. Within each of those directories, a directory for each unique hash of the PDB, which has the PDB file inside it. This structure eliminates the need to search the entire directory, it just looks for a directory of < binary > \ < hash >.

**Comment by yharaskrik42** [comment]
@KenSykes - Yeah, putting "srv*" before any symbol directory solves the problem for me. Thank you!
